doc/elements-tx-format.md

```diff
@@ -0,0 +1,570 @@
+
+This document describes the format used to serialize Elements transactions. Once a transaction has been converted into this raw, serialized form, it can be broadcast across the network.
+
+This document requires familiarity with the architecture of Bitcoin and Elements, as well the various parts that are expected in an Elements transaction (inputs, outputs, assets, pegin data, etc.). For more information on those, please refer to the [Elements code tutorial](https://elementsproject.org/elements-code-tutorial/overview).
+
+### Data Types
+
+*Note:* The encoding column in the following tables corresponds to the way that the fields are serialized. In Bitcoin, numbers and hashes are serialized in little-endian, and this same notion carries over to Elements.
+
+*Note:* Values are listed in multiples of 100,000,000 (this corresponds directly to how Bitcoin encodes value, where 1 BTC is divisible into 100,000,000 Satoshis). For example, a `ConfidentialAmount` representing 7 inflation keys would be encoded as 700,000,000.
+
+#### Transaction
+
+|Field|Required|Size|Data Type|Encoding|Notes| 
+|-----|--------|-----|---------|--------|-----|
+| Version | Yes | 4 bytes | `int32_t` | Little-endian | Transaction version number. Currently version 2 (see [BIP 68](https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki#specification)). | |
+| Flags | Yes | 1 byte | `unsigned char` | | Whether the transaction contains a witness. Either 0 (no witness) or 1. |
+| Num Inputs | Yes | Varies | `VarInt` | | Number of inputs to the transaction. |
+| Inputs | Yes | Varies | `Vector<TxIn>` | | |
+| Num Inputs | Yes | Varies | `VarInt` | | Number of outputs from the transaction. |
+| Outputs | Yes | Varies | `Vector<TxOut>` | | |
+| Locktime | Yes | 4 bytes | `uint32_t` | Little-endian | See [BIP 113](https://github.com/bitcoin/bips/blob/master/bip-0113.mediawiki). |
+| Witness | Only if flags is 1 | Varies | `Witness` | | See [BIP 141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki). Note that Elements witnesses contain more data than Bitcoin witnesses. This extra data is described further below. |
+
+Serialized format: `[Version][Flags][Num Inputs][Inputs][Num Outputs][Outputs][Locktime][Witness]`
+
+This is the overarching structure of a serialized transaction. The rest of this document contains further details on specific parts, as well as examples.
+
+#### TxIn
+
+|Field|Required|Size|Data Type|Encoding|Notes| 
+|-----|--------|-----|---------|--------|-----|
+| Output Hash (TXID) | Yes | 32 bytes | `uint256` | Little-endian | Bytes of the output hash are encoded in little-endian. **NB**: this means that they appear backwards when serialized. |
+| Output Index | Yes | 4 bytes | `uint32_t` | Little-endian | **Input is a coinbase**: `0xffffffff`<br><br>The two most significant bits are reserved for flags.<br><br>**Input is a pegin:** second most significant bit is 1.<br><br>**Input has an asset issuance:** most significant bit is 1. |
```

Comment:
what happens if both bits are 1?

Reply:
I think it's self-evident that an asset issuance and a pegin are mutually exclusive, from a logical perspective. 

I'd prefer not to enumerate every single thing that could be invalid, because that would make the document much longer and more confusing to read.

Reply:
I think it's self-evident that an asset issuance and a pegin are mutually exclusive, from a logical perspective.

I'd prefer not to enumerate every single thing that could be invalid, because that would make the document much longer and more confusing to read.

I could add a link to the actual transaction validation code [here](https://github.com/ElementsProject/elements/blob/a49c01693b27148cdf188c85b89b83fe0b598ba1/src/consensus/tx_check.cpp#L10), but I'm not sure how helpful that will be.

Reply:
As far as I am aware you can pegin and issue an asset at the same time. I've never tried it though.